reporter/src/main/kotlin/reporters/fossid/FossIdReporter.kt

```diff
@@ -0,0 +1,90 @@
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.reporter.reporters.fossid
+
+import java.io.File
+
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.runBlocking
+
+import org.ossreviewtoolkit.clients.fossid.FossIdRestService
+import org.ossreviewtoolkit.clients.fossid.generateReport
+import org.ossreviewtoolkit.clients.fossid.model.report.ReportType
+import org.ossreviewtoolkit.clients.fossid.model.report.SelectionType
+import org.ossreviewtoolkit.reporter.Reporter
+import org.ossreviewtoolkit.reporter.ReporterInput
+import org.ossreviewtoolkit.utils.core.log
+
+class FossIdReporter : Reporter {
+    companion object {
+        /** Name of the configuration property for the server URL. */
+        const val SERVER_URL_PROPERTY = "serverUrl"
+
+        /** Name of the configuration property for the API key. */
+        const val API_KEY_PROPERTY = "apiKey"
+
+        /** Name of the configuration property for the user name. */
+        const val USER_PROPERTY = "user"
+
+        /** Name of the configuration property for the report type. Default is [ReportType.HTML_DYNAMIC]. */
+        const val REPORT_TYPE_PROPERTY = "reportType"
+
+        /**
+         * Name of the configuration property for the selection type.
+         * Default is [SelectionType.INCLUDE_ALL_LICENSES].
+         */
+        const val SELECTION_TYPE_PROPERTY = "selectionType"
+
+        /**
+         * Name of key in [ScanResult.additionalData] containing the scancode.
+         * TODO: It should be unified with [FossId.SCAN_CODE_KEY], without creating a dependency between scanner and
+         * reporter.
+         */
+        const val SCAN_CODE_KEY = "scancode"
+    }
+
+    override val reporterName = "FossId"
+
+    override fun generateReport(input: ReporterInput, outputDir: File, options: Map<String, String>): List<File> {
+        val serverUrl = options[SERVER_URL_PROPERTY]
+            ?: throw IllegalArgumentException("No FossID server URL configuration found.")
+        val apiKey = options[API_KEY_PROPERTY]
+            ?: throw IllegalArgumentException("No FossID API Key configuration found.")
+        val user = options[USER_PROPERTY]
+            ?: throw IllegalArgumentException("No FossID User configuration found.")
+        val reportType = options[REPORT_TYPE_PROPERTY]?.let { ReportType.valueOf(it) } ?: ReportType.HTML_DYNAMIC
+        val selectionType =
+            options[SELECTION_TYPE_PROPERTY]?.let { SelectionType.valueOf(it) } ?: SelectionType.INCLUDE_ALL_LICENSES
```

Comment:
These would crash if the options were non-null but invalid values. So maybe better:

    val reportType = runCatching { ReportType.valueOf(options[REPORT_TYPE_PROPERTY]) }
        .getOrDefault(ReportType.HTML_DYNAMIC)

Reply:
hmmm this code is not working. It should be:
```
val reportType = runCatching {           
           ReportType.valueOf(options[REPORT_TYPE_PROPERTY].orEmpty())
}.getOrDefault(ReportType.HTML_DYNAMIC)
```
however, although you have a point, I don't like this solution: if the parameter is null, it will nevertheless be parsed with `valueOf` throw an exception, then return the default value.
```
val reportType =  options[REPORT_TYPE_PROPERTY]?.let { 
            kotlin.runCatching { 
                ReportType.valueOf(it)}.getOrNull()
        } ?: ReportType.HTML_DYNAMIC
```

Reply:
> if the parameter is null, it will nevertheless be parsed with `valueOf` throw an exception, then return the default value.

But isn't this exactly the behavior that we want?

> hmmm this code is not working.

That's odd, as `valueOf()` is documented to throw an NPE if the argument is *null*, so the wrapping `runCatching` *should* catch it...

Reply:
> But isn't this exactly the behavior that we want?

You didn't understand what I meant. Yes, this is what we want but trying to parse a _null_ value is just wasting unnecessary CPU cycle. My code skips the call to `valueOf` if the option is `null`.

> > hmmm this code is not working.
> 
> That's odd, as `valueOf()` is documented to throw an NPE if the argument is _null_, so the wrapping `runCatching` _should_ catch it...

I could not get an online confirmation but it seems `valueOf` of Kotlin is NOT the `valueOf` of java. Therefore the parameter is of type `String` not `String?`.

Reply:
> You didn't understand what I meant.

Indeed 😁 Because I thought by "hmmm this code is not working" you mean that the code is not behaving as expected at runtime. But now I understand you mean that it does not compile.